### PR TITLE
Editor: Display trash button only if user has delete capability

### LINKED ISSUE #14828: "Move to Trash" button doesn't respect 'delete_post' capability
**Describe the bug**

When a user cannot delete a post (as defined by the `delete_post` meta capability), the "Move to Trash" button still displays. My expectation is that the "Move to Trash" button wouldn't display.

**To reproduce**

1. Add the following code snippet to a mu-plugin:

```php
add_filter(
	'map_meta_cap',
	function( $caps, $cap, $user_id, $args ) {
		if ( 'delete_post' === $cap ) {
			$caps = [ 'do_not_allow' ];
		}
		return $caps;
	},
	10,
	4
);
```

2. Active the Classic Editor plugin and view an existing post. The 'Move to Trash' link isn't present:

image

3. Deactivate the Classic Editor plugin and open the same post in the Block Editor. The 'Move to Trash' link is present, even though the REST API errors when you attempt to trash the post:

image

**Suggested fix**

Similar to the work done for #6361, we'll need to create some new `wp:action-delete` JSON Hyper Schema `targetSchema` response data for posts.

After this, the `PostTrash` component in `@wordpress/editor` will need to be updated to respect the presence of the `wp:action-delete` metadata.

### PR DESCRIPTION
**Pull request by aduth**
Fixes #14828 

This pull request seeks to resolve an issue where the "Move to Trash" button is shown even if the user does not have the correct capabilities to be able to trash the post.

**Implementation notes:**

I followed the suggested implementation from @talldan at https://github.com/wordPress/gutenberg/issues/14828#issuecomment-480149595 . I could not encounter the same issued described from the comment there with respect to a 403 (Forbidden) response. @talldan could you check to see if the issue is present here, and if so, elaborate on the instructions to reproduce it?

**Testing Instructions:**

Repeat Steps to Reproduce from #14828, verifying that the Move to Trash button is only shown when the user has capabilities to delete the post.

### CODE REVIEW

**Comment by aduth** [comment]
There appear to be some legitimate failures in the end-to-end suite, specifically for the template tests, in encountering a 404 error from one of its requests.

>["Failed to load resource: the server responded with a status of 404 (Not Found)"]

I'll plan to take a look.

File: packages/editor/src/components/post-trash/check.js [hunk 1]
```diff
@@ -13,8 +13,12 @@ function PostTrashCheck( { isNew, postId, children } ) {
 
 export default withSelect( ( select ) => {
 	const { isEditedPostNew, getCurrentPostId } = select( 'core/editor' );
+	const { canUser } = select( 'core' );
+	const postId = getCurrentPostId();
+
 	return {
 		isNew: isEditedPostNew(),
-		postId: getCurrentPostId(),
+		canUserTrash: !! postId && canUser( 'delete', 'posts', postId ),
```
Comment (talldan) [inline packages/editor/src/components/post-trash/check.js]: Should the postType be hard-coded here? I'm seeing a 404 from the request that the resolver makes when trying to edit a reusable block. That might also be the cause of the test failures.

IIRC, the `rest_base` from the `postType` is the value that should be used.

**Review by talldan — CHANGES_REQUESTED** [review]
I think the tests have caught a problem related to the postType, see the comment below.

> @talldan could you check to see if the issue is present here, and if so, elaborate on the instructions to reproduce it?

It looks like this only happens for drafts. My steps were:
1. Create a draft as an admin and take note of the post id
2. Login as a contributor and open up the editor
3. In the console run `wp.data.select( 'core' ).canUser( 'delete', 'posts', postId );` using the postId from step 1.
4. Observe that the request returns a 403.

Here's my console output:
Screen Shot 2019-05-01 at 10 30 17 am

`8` is a draft and `6` is published. 

In real world usage this isn't much of an issue, a contributor can't load up one of these posts in the editor and the selector still returns a falsey.

**Comment by aduth** [comment]
I forget why I had never circled back here. Apologies for allowing your review to linger without response @talldan.

I don't know if you recall what you were expecting, but are you thinking the problem is in how (a) a network error is logged (might be out of our control and default Chrome behavior) and (b) the selector returns `undefined` and not exclusively either `true` or `false` (a boolean value)?

File: packages/editor/src/components/post-trash/check.js (hunk 1 above)
Comment (aduth) [reply packages/editor/src/components/post-trash/check.js]: > Should the postType be hard-coded here? I'm seeing a 404 from the request that the resolver makes when trying to edit a reusable block. That might also be the cause of the test failures.
> 
> IIRC, the `rest_base` from the `postType` is the value that should be used.

Yep, I think you're correct on both accounts.

**Comment by aduth** [comment]
Looking at the implementation, the conditional rendering probably doesn't need to happen in `PostTrash`. It should be up to the render usage (notably, `@wordpress/edit-post`'s implementation) to decide if and how to use the `PostTrashCheck`. Otherwise, it seems there's no point in having `PostTrashCheck` to begin with.

**Comment by talldan** [comment]
@aduth Struggling to remember all the details about this one, sorry!

I supposed that the `canUser` resolver doesn't specifically check for a 403 and return a false might be what I was thinking. But it does return `falsey` so not so much of an issue.

**Comment by aduth** [comment]
> I supposed that the `canUser` resolver doesn't specifically check for a 403 and return a false might be what I was thinking. But it does return `falsey` so not so much of an issue.

I think `false` would be the expected return value. Whether it's being handled? Hard to tell at the moment. I see it does a `try` / `catch` on the request itself:

https://github.com/WordPress/gutenberg/blob/56ff721e76b6412d45eae43df10dcf6524b4f83b/packages/core-data/src/resolvers.js#L159-L173

But I'd want to confirm:

- If `apiFetch` actually throws based on status codes. I know this is a pain point with `window.fetch`, that non-success HTTP status codes are still treated as success for the promise itself (i.e. an error is not thrown, and it's up to the developer to confirm the status code)
- Whether, even if the request does not throw, if the [remaining logic](https://github.com/WordPress/gutenberg/blob/56ff721e76b6412d45eae43df10dcf6524b4f83b/packages/core-data/src/resolvers.js#L175-L188) would be sufficient to accommodate this, even for a failed request.

**Comment by talldan** [comment]
This has been resolved by https://github.com/WordPress/gutenberg/pull/23174

**Pull request closed without merging**